verify.js

```diff
@@ -110,18 +111,39 @@ module.exports = function (jwtString, secretOrPublicKey, options, callback) {
       options.algorithms = ['none'];
     }
 
-    if (!options.algorithms) {
-      options.algorithms = secretOrPublicKey.toString().includes('BEGIN CERTIFICATE') ||
-        secretOrPublicKey.toString().includes('BEGIN PUBLIC KEY') ? PUB_KEY_ALGS :
-        secretOrPublicKey.toString().includes('BEGIN RSA PUBLIC KEY') ? RSA_KEY_ALGS : HS_ALGS;
+    if (secretOrPublicKey != null && !(secretOrPublicKey instanceof KeyObject)) {
+      try {
+        secretOrPublicKey = createPublicKey(secretOrPublicKey);
+      } catch (_) {
+        try {
+          secretOrPublicKey = createSecretKey(typeof secretOrPublicKey === 'string' ? Buffer.from(secretOrPublicKey) : secretOrPublicKey);
+        } catch (_) {
+          return done(new JsonWebTokenError('secretOrPublicKey is not valid key material'))
+        }
+      }
+    }
 
+    if (!options.algorithms) {
+      if (secretOrPublicKey.type === 'secret') {
+        options.algorithms = HS_ALGS;
+      } else if (secretOrPublicKey.asymmetricKeyType === 'rsa') {
+        options.algorithms = RSA_KEY_ALGS
+      } else {
+        options.algorithms = PUB_KEY_ALGS
+      }
```

Comment:
This is a slight change from previous behaviour: if we had a PEM with `BEGIN RSA PUBLIC KEY`, we knew we could narrow down to RSA only, otherwise for `BEGIN PUBLIC KEY` we allow all public algs.

The former case could have `asymmetricKeyType === 'pub-rsa'` so we end up allowing slightly more than before. 

IMO let's pick up RSA-PSS here, and also handle ECC while we're here for consistency (technically a breaking change for EC though so we need to document it):

```
      if (secretOrPublicKey.type === 'secret') {
        options.algorithms = HS_ALGS;
      } else if (secretOrPublicKey.asymmetricKeyType === 'rsa-pss' || secretOrPublicKey.asymmetricKeyType === 'rsa-pss') {
        options.algorithms = RSA_KEY_ALGS;
      } if (secretOrPublicKey.asymmetricKeyType === 'ec') {
        options.algorithms = EC_KEY_ALGS;
      }
       else {
        options.algorithms = PUB_KEY_ALGS;
      }
```

Reply:
It'd be great to add some explanation of this in the README as well - we don't specify what happens if `algorithms` is undefined

Reply:
Done. We'll add the change to our migration guide